server/src/main/java/org/elasticsearch/index/query/ExistsQueryBuilder.java

```diff
@@ -164,22 +164,28 @@ public static Query newFilter(QueryShardContext context, String fieldPattern) {
         return new ConstantScoreQuery(boolFilterBuilder.build());
     }
 
-    private static Query newLegacyExistsQuery(Collection<String> fields) {
+    private static Query newLegacyExistsQuery(QueryShardContext context, Collection<String> fields) {
         // We create TermsQuery directly here rather than using FieldNamesFieldType.termsQuery()
         // so we don't end up with deprecation warnings
         if (fields.size() == 1) {
-            Query filter = new TermQuery(new Term(FieldNamesFieldMapper.NAME, fields.iterator().next()));
+            Query filter = newLegacyTermQuery(context, fields.iterator().next());
             return new ConstantScoreQuery(filter);
         }
 
         BooleanQuery.Builder boolFilterBuilder = new BooleanQuery.Builder();
         for (String field : fields) {
-            Query filter = new TermQuery(new Term(FieldNamesFieldMapper.NAME, field));
+            Query filter = newLegacyTermQuery(context, field);
             boolFilterBuilder.add(filter, BooleanClause.Occur.SHOULD);
         }
         return new ConstantScoreQuery(boolFilterBuilder.build());
     }
 
+    private static Query newLegacyTermQuery(QueryShardContext context, String field) {
+        MappedFieldType fieldType = context.fieldMapper(field);
```

Comment:
I've been a bit concerned that this requirement (to fetch the name of the concrete field type) can be 'trappy' for other developers. I don't see a way to avoid this, though.

Reply:
It looks ok to me